### PR TITLE
backend/repository: pull layers in parallel

### PR DESCRIPTION
**Pull request by dgonyeo**
To reduce the time it takes to fetch docker images, this commit modifies
the backends to pull each image layer concurrently. As a result,
github.com/coreos/pkg/progressutil is used in place of ioprogress to
draw the progress bars.

### CODE REVIEW

File: lib/internal/backend/repository/repository1.go [hunk 1]
```diff
@@ -62,36 +62,74 @@ func (rb *RepositoryBackend) getImageInfoV1(dockerURL *types.ParsedDockerURL) ([
...
 }
 
-func (rb *RepositoryBackend) buildACIV1(layerNumber int, layerID string, dockerURL *types.ParsedDockerURL, outputDir string, tmpBaseDir string, curPwl []string, compression common.Compression) (string, *schema.ImageManifest, error) {
-	tmpDir, err := ioutil.TempDir(tmpBaseDir, "docker2aci-")
-	if err != nil {
-		return "", nil, fmt.Errorf("error creating dir: %v", err)
-	}
-	defer os.RemoveAll(tmpDir)
+func (rb *RepositoryBackend) buildACIV1(layerIDs []string, dockerURL *types.ParsedDockerURL, outputDir string, tmpBaseDir string, compression common.Compression) ([]string, []*schema.ImageManifest, error) {
+	//func (rb *RepositoryBackend) buildACIV1(layerNumber int, layerID string, dockerURL *types.ParsedDockerURL, outputDir string, tmpBaseDir string, curPwl []string, compression common.Compression) (string, *schema.ImageManifest, error) {
```
Comment (jonboulle) [inline lib/internal/backend/repository/repository1.go]: meep meep

**Comment by jonboulle** [comment]
@dgonyeo afaict this is  a legit test failure - file is not getting white outed as it should. Can you take a look?

**Comment by jonboulle** [comment]
bump

**Comment by dgonyeo** [comment]
After a couple of days of banging my head against the wall, I can't figure out why that test is failing. I didn't expect to have changed the logic related to whiteouts, and I pass test-pwl and fail test-whiteouts, yet I don't understand the differences between the two tests.

Would @iaguis have the time to give me a hand at figuring out why I'm failing test-whiteouts?

File: lib/docker2aci.go [hunk 2]
```diff
@@ -124,31 +124,25 @@ func convertReal(backend internal.Docker2ACIBackend, dockerURL string, squash bo
...
-			return nil, fmt.Errorf("error building layer: %v", err)
-		}
+	aciLayerPaths, aciManifests, err := backend.BuildACI(ancestry, parsedDockerURL, layersOutputDir, tmpDir, layerCompression)
+	if err != nil {
+		return nil, err
+	}
 
-		key, err := conversionStore.WriteACI(aciPath)
+	var images acirenderer.Images
+	for i, aciLayerPath := range aciLayerPaths {
```
Comment (iaguis) [inline lib/docker2aci.go]: Although `aciLayerPath` has the right order (from layer `len(ancestry) - 1` to 0) the index `i` is ascending in this loop.

You need to change the `uint16(i)` below to `uint16(len(aciLayerPaths) - 1 - i)`.

Comment (iaguis) [inline lib/docker2aci.go]: https://semaphoreci.com/iaguis/docker2aci/branches/iaguis-parallel-pulls/builds/3

**Comment by iaguis** [comment]
Let's continue on #158. Thanks a lot for this!

**Pull request closed without merging**